### DISCUSSION TITLE
Portable mode for Zed

### DISCUSSION THREAD

**Discussion by adament** [zed-industries/zed#21103] [Show and tell]
In [the branch portable_mode on adament/zed](https://github.com/adament/zed/tree/portable_mode) I have made some small modifications to the paths crate to enable a portable mode for Zed. The portable mode, when active, means that Zed stores all config, support, temp, logs, and crashes in a user specified directory rather than the system default folders. This is useful in a couple of cases:
1. Quickly testing alternate configurations / clean setups / extensions without modifying the global default Zed state.
2. Keeping a zed executable including all configuration and state on a movable media such as a USB stick or network drive. In such a way that the user can easily transfer between computers and bring their editor setup along with the editor.
3. Some corporate Windows environments prohibit installing extensions to AppData

In the current implementation I have gone for simple and explicit behaviour. But this is meant as a basis for a discussion rather than the only true way. I have tried to lay out my rationale for the current design below, but I hope to hear whether other people consider such a feature useful and if we can come up with a better design.

# Activating portable mode
Currently the implementation uses the environment variable  `ZED_PORTABLE_DIR` to activate portable mode and specify the root folder to store everything in. This has the disadvantage that the user must set this environment variable to activate portable mode. Which means they need some wrapper script or executable to start Zed in use case 2. Similarly for package managers like Scoop on Windows which generally uses portable mode applications.

As an alternative Visual Studio Code activates portable mode by detecting whether a folder `data` exists next to the executable. While this is a neat solution I think it has more steps which could fail and thus need to be considered: Determine the location of the executable, detect whether a folder exists next to it, and handle the case of Application Bundles on macos.

A third option could be to simply make portable mode a compile time directive, but that means separate executables for portable mode and regular installation.

# Temporary files
The current implementation also stores temporary files under the portable folder, this is in an attempt to make the portable version as self-contained as possible. However this might be considered surprising and come with performance impacts in the case the portable mode folder is a network drive or usb stick.

Visual Studio Code uses the system temp folder by default but detects if a `tmp`  folder exists inside the portable mode folder, and if it does uses the `data\tmp` folder instead.

**Comment by Igonato** [comment]
Created this issue the other day https://github.com/zed-industries/zed/issues/21131 
Linking for relevance.

**Comment by SomeoneToIgnore** [comment]
https://github.com/zed-industries/zed/discussions/44116

**Comment by Dreamoochy** [comment]
What about just renaming zed executable? I've seen that approach before (IIRC, with avidemux). After start, Zed could get its process executable name and check whether it contains "portable". Maybe there's no big difference between this and checking a `data` folder, but at least `zed_portable.exe` would immediately tell you what to expect.

**Comment by EdwinHolley** [comment]
A environment variable does not make an app portable to me.  It makes it so it does not use Windows methods of storing info/config/etc. Which is half of what I consider for a portable app, the second part is to be able to move it from computer to computer.  Which means bring up a command prompt or an app to set an environment variable.  To me, the setup should set a configuration option or the like and all data be stored in subfolders of where the application is being stored, in a standard format for easy reading.  If necessary these could be moved by changing default settings or even offering the option during setup.  I use portable apps exclusively when I am at other peoples computers.

My $.02
Ed
